core/silkworm/common/endian.cpp

```diff
@@ -0,0 +1,69 @@
+/*
+   Copyright 2021 The Silkworm Authors
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+#include "endian.hpp"
+
+#include <silkworm/common/util.hpp>
+
+namespace silkworm::endian {
+
+Bytes to_big_compact(const uint64_t value) {
+    uint8_t full_be[sizeof(uint64_t)];
+    store_big_u64(&full_be[0], value);
+    return Bytes{zeroless_view(full_be)};
+}
+
+Bytes to_big_compact(const intx::uint256& value) {
+    uint8_t full_be[sizeof(intx::uint256)];
+    intx::be::store(full_be, value);
+    return Bytes{zeroless_view(full_be)};
+}
+
+template <typename T>
+static std::optional<T> from_big_compact(const ByteView& data, bool allow_leading_zeros) {
+    if (data.length() > sizeof(T)) {
+        return std::nullopt;
+    }
+
+    T x{0};
+
+    if (data.empty()) {
+        return x;
+    }
```

Comment:
Probably worth gather the sanity checks which return std::nullopt in a single place
```suggestion
static std::optional<T> from_big_compact(const ByteView& data, bool allow_leading_zeros) {

    if (data.length() > sizeof(T) ||                               // Exceeds size of T
        (!data.empty() && (!allow_leading_zeros && data[0] == 0))  // Leading zeroes not allowed
    ) {
        return std::nullopt;
    }

    T x{0};
    if (data.empty()) {
        return x;
    }

```

Reply:
Hmm. I find the separate checks more readable.

Reply:
Yes I know but the _less readable_ form saves the creation of instance of T which is eventually wasted cause we return std::nullopt